From 8913432378008561ea697a94af8b25d4f0e6322b Mon Sep 17 00:00:00 2001
From: Hongting Ting <hongting.dong@nxp.com>
Date: Thu, 25 Jul 2024 16:45:40 +0800
Subject: [PATCH 07/19] arm64: dts: add imx93-11x11-frdm-aud-hat dts

Add CS42448 CODEC node
Add XCVR node
Add micfil node to record audio

Signed-off-by: Hongting Ting <hongting.dong@nxp.com>
Signed-off-by: Steven <zhao.yang_1@nxp.com>
Signed-off-by: Lei Xu <Lei.Xu@nxp.com>
---
 arch/arm64/boot/dts/freescale/Makefile        |   3 +-
 .../freescale/imx93-11x11-frdm-aud-hat.dts    | 316 ++++++++++++++++++
 2 files changed, 318 insertions(+), 1 deletion(-)
 create mode 100644 arch/arm64/boot/dts/freescale/imx93-11x11-frdm-aud-hat.dts

diff --git a/arch/arm64/boot/dts/freescale/Makefile b/arch/arm64/boot/dts/freescale/Makefile
index 68f81cdc2308..9c243f7798be 100644
--- a/arch/arm64/boot/dts/freescale/Makefile
+++ b/arch/arm64/boot/dts/freescale/Makefile
@@ -385,7 +385,8 @@ dtb-$(CONFIG_ARCH_MXC) += imx93-11x11-evk.dtb \
 			  imx93-11x11-evk-iw612-otbr.dtb \
 			  imx93-11x11-evk-rpmsg.dtb imx93-11x11-evk-rpmsg-lpv.dtb \
 			  imx93-11x11-frdm.dtb imx93-11x11-frdm-mt9m114.dtb \
-			  imx93-11x11-frdm-dsi.dtb imx93-11x11-frdm-tianma-wvga-panel.dtb
+			  imx93-11x11-frdm-dsi.dtb imx93-11x11-frdm-tianma-wvga-panel.dtb \
+			  imx93-11x11-frdm-aud-hat.dtb
 
 dtb-$(CONFIG_ARCH_MXC) += imx91-11x11-evk.dtb \
 			  imx91-11x11-evk-flexspi-m2.dtb imx91-11x11-evk-flexspi-nand-m2.dtb \
diff --git a/arch/arm64/boot/dts/freescale/imx93-11x11-frdm-aud-hat.dts b/arch/arm64/boot/dts/freescale/imx93-11x11-frdm-aud-hat.dts
new file mode 100644
index 000000000000..8e6386df97c7
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx93-11x11-frdm-aud-hat.dts
@@ -0,0 +1,316 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
+/*
+ * Copyright 2024 NXP
+ */
+
+#include "imx93-11x11-frdm.dts"
+
+/ {
+	mic_leds {
+		compatible = "gpio-leds";
+		mic0 {
+			label = "mic0";
+			gpios = <&pca9555 5 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+		mic1 {
+			label = "mic1";
+			gpios = <&pca9555 7 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+		mic2 {
+			label = "mic2";
+			gpios = <&pca9555 6 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+		mic3 {
+			label = "mic3";
+			gpios = <&pca9555 2 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+		mic4 {
+			label = "mic4";
+			gpios = <&pca9555 1 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+		mic5 {
+			label = "mic5";
+			gpios = <&pca9555 0 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+		mic6 {
+			label = "mic6";
+			gpios = <&pca9555 4 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+		mic7 {
+			label = "mic7";
+			gpios = <&pca9555 3 GPIO_ACTIVE_LOW>;
+			default-state = "off";
+		};
+	};
+
+	sw_keys {
+		compatible = "gpio-keys";
+
+		sw4: volume_down {
+			label = "Volume Down";
+			gpios = <&pca9555 15 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_VOLUMEDOWN>;
+			interrupt-parent = <&pca9555>;
+			interrupts = <15 IRQ_TYPE_LEVEL_LOW>;
+		};
+
+		sw3: volume_up {
+			label = "Volume Up";
+			gpios = <&pca9555 14 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_VOLUMEUP>;
+			interrupt-parent = <&pca9555>;
+			interrupts = <14 IRQ_TYPE_LEVEL_LOW>;
+		};
+
+		sw2: volume_mute {
+			label = "Volume Mute";
+			gpios = <&pca9555 13 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_MUTE>;
+			interrupt-parent = <&pca9555>;
+			interrupts = <13 IRQ_TYPE_LEVEL_LOW>;
+		};
+
+		sw1: key_act {
+			label = "Key Act";
+			gpios = <&pca9555 12 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_F9>;
+			interrupt-parent = <&pca9555>;
+			interrupts = <12 IRQ_TYPE_LEVEL_LOW>;
+		};
+	};
+
+	exp_3v3_switch: regulator-3v3-switch {
+		compatible = "regulator-fixed";
+		regulator-name = "exp-3v3-switch";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		gpio = <&pcal6524 2 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		regulator-always-on;
+	};
+
+	exp_5v_switch: regulator-5v-switch {
+		compatible = "regulator-fixed";
+		regulator-name = "exp-5v-switch";
+		regulator-min-microvolt = <5000000>;
+		regulator-max-microvolt = <5000000>;
+		gpio = <&pcal6524 8 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		regulator-always-on;
+	};
+
+	reg_audio_pwr: regulator-audio-pwr {
+		compatible = "regulator-fixed";
+		pinctrl-names = "default";
+		pinctrl-0 = <&pinctrl_audio_pwr>;
+		regulator-name = "cs42448-supply";
+		regulator-min-microvolt = <3300000>;
+		regulator-max-microvolt = <3300000>;
+		gpio = <&gpio2 1 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+		regulator-always-on;
+	};
+
+	sound-xcvr {
+		compatible = "fsl,imx-audio-card";
+		model = "imx-audio-xcvr";
+		pri-dai-link {
+			link-name = "XCVR PCM";
+			cpu {
+				sound-dai = <&xcvr>;
+			};
+		};
+	};
+
+	sound-micfil {
+		compatible = "fsl,imx-audio-card";
+		model = "micfil-audio";
+		pri-dai-link {
+			link-name = "micfil hifi";
+			format = "i2s";
+			cpu {
+				sound-dai = <&micfil>;
+			};
+		};
+	};
+
+	sound-cs42448 {
+		compatible = "fsl,imx-audio-card";
+		model = "imx-cs42448";
+		status = "okay";
+		pri-dai-link {
+			link-name = "cs42448";
+			format = "dsp_a";
+			dai-tdm-slot-num = <8>;
+			dai-tdm-slot-width = <32>;
+			fsl,mclk-equal-bclk;
+			cpu {
+				sound-dai = <&sai3>;
+			};
+			codec {
+				sound-dai = <&cs42448>;
+			};
+		};
+	};
+};
+
+&sai3 {
+	#sound-dai-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_sai3audhat>;
+	clocks = <&clk IMX93_CLK_SAI3_IPG>, <&clk IMX93_CLK_DUMMY>,
+		<&clk IMX93_CLK_SAI3_GATE>,
+		<&clk IMX93_CLK_DUMMY>, <&clk IMX93_CLK_DUMMY>,
+		<&clk IMX93_CLK_AUDIO_PLL>;
+	clock-names = "bus", "mclk0", "mclk1", "mclk2", "mclk3", "pll8k";
+	assigned-clocks = <&clk IMX93_CLK_SAI3>;
+	assigned-clock-parents = <&clk IMX93_CLK_AUDIO_PLL>;
+	assigned-clock-rates = <12288000>;
+	fsl,sai-asynchronous;
+	fsl,sai-mclk-direction-output;
+	status = "okay";
+};
+
+&flexcan2 {
+	status = "disabled";
+};
+
+&xcvr {
+	#sound-dai-cells = <0>;
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&pinctrl_spdif>;
+	pinctrl-1 = <&pinctrl_spdif_sleep>;
+	clocks = <&clk IMX93_CLK_BUS_WAKEUP>,
+		<&clk IMX93_CLK_SPDIF_GATE>,
+		<&clk IMX93_CLK_DUMMY>,
+		<&clk IMX93_CLK_AUD_XCVR_GATE>,
+		<&clk IMX93_CLK_AUDIO_PLL>;
+	clock-names = "ipg", "phy", "spba", "pll_ipg", "pll8k";
+	assigned-clocks = <&clk IMX93_CLK_SPDIF>,
+			 <&clk IMX93_CLK_AUDIO_XCVR>;
+	assigned-clock-parents = <&clk IMX93_CLK_AUDIO_PLL>,
+			 <&clk IMX93_CLK_SYS_PLL_PFD1_DIV2>;
+	assigned-clock-rates = <12288000>, <200000000>;
+	status = "okay";
+};
+
+&micfil {
+	#sound-dai-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_pdm>;
+	assigned-clocks = <&clk IMX93_CLK_PDM>;
+	assigned-clock-parents = <&clk IMX93_CLK_AUDIO_PLL>;
+	assigned-clock-rates = <49152000>;
+	status = "okay";
+};
+
+&lpi2c4 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	clock-frequency = <400000>;
+	dmas = <&edma2 10 0 0>, <&edma2 11 0 1>;
+	dma-names = "tx","rx";
+	pinctrl-names = "default", "sleep";
+	pinctrl-0 = <&pinctrl_lpi2c4>;
+	pinctrl-1 = <&pinctrl_lpi2c4>;
+	status = "okay";
+
+	pca9555: gpio@21 {
+		compatible = "nxp,pca9555";
+		pinctrl-names = "default";
+		pinctrl-0 = <&pinctrl_pushbutton_irq>;
+		reg = <0x21>;
+		gpio-controller;
+		#gpio-cells = <2>;
+		interrupt-controller;
+		#interrupt-cells = <2>;
+		interrupt-parent = <&gpio2>;
+		interrupts = <27 IRQ_TYPE_LEVEL_LOW>;
+		vcc-supply = <&exp_3v3_switch>;
+		status = "okay";
+	};
+
+	cs42448: cs42448@48 {
+		#sound-dai-cells = <0>;
+		compatible = "cirrus,cs42448";
+		reg = <0x48>;
+		clocks = <&clk IMX93_CLK_SAI3_GATE>;
+		clock-names = "mclk";
+		VA-supply = <&reg_audio_pwr>;
+		VD-supply = <&reg_audio_pwr>;
+		VLS-supply = <&reg_audio_pwr>;
+		VLC-supply = <&reg_audio_pwr>;
+		reset-gpio = <&gpio2 25 GPIO_ACTIVE_LOW>;
+		status = "okay";
+	};
+};
+
+&iomuxc {
+	pinctrl_swpdm_mute_irq: swpdm_mute_grp {
+		fsl,pins = <
+			MX93_PAD_GPIO_IO00__GPIO2_IO00	0x31e
+		>;
+	};
+
+	pinctrl_pushbutton_irq: pushbutton_grp {
+		fsl,pins = <
+			MX93_PAD_GPIO_IO27__GPIO2_IO27	0x31e
+		>;
+	};
+
+	pinctrl_pdm: pdmgrp {
+		fsl,pins = <
+			MX93_PAD_GPIO_IO04__PDM_CLK		0x31e
+			MX93_PAD_GPIO_IO05__PDM_BIT_STREAM00	0x31e
+			MX93_PAD_GPIO_IO06__PDM_BIT_STREAM01	0x31e
+			MX93_PAD_GPIO_IO12__PDM_BIT_STREAM02	0x31e
+			MX93_PAD_GPIO_IO13__PDM_BIT_STREAM03	0x31e
+		>;
+	};
+
+	pinctrl_lpi2c4: lpi2c4grp {
+		fsl,pins = <
+			MX93_PAD_GPIO_IO03__LPI2C4_SCL		0x40000b9e
+			MX93_PAD_GPIO_IO02__LPI2C4_SDA		0x40000b9e
+		>;
+	};
+
+	pinctrl_sai3audhat: sai3audhat {
+		fsl,pins = <
+			MX93_PAD_GPIO_IO17__SAI3_MCLK		0x31e
+			MX93_PAD_GPIO_IO26__SAI3_TX_SYNC	0x31e
+			MX93_PAD_GPIO_IO16__SAI3_TX_BCLK	0x31e
+			MX93_PAD_GPIO_IO19__SAI3_RX_SYNC	0x31e
+			MX93_PAD_GPIO_IO18__SAI3_RX_BCLK	0x31e
+			MX93_PAD_GPIO_IO21__SAI3_TX_DATA00	0x31e
+			MX93_PAD_GPIO_IO20__SAI3_RX_DATA00	0x31e
+		>;
+	};
+
+	pinctrl_audio_pwr: audiopwrgrp {
+		fsl,pins = <
+			MX93_PAD_GPIO_IO01__GPIO2_IO01		0x31e
+		>;
+	};
+
+	pinctrl_spdif: spdifgrp {
+		fsl,pins = <
+			MX93_PAD_GPIO_IO22__SPDIF_IN		0x31e
+			MX93_PAD_GPIO_IO23__SPDIF_OUT		0x31e
+		>;
+	};
+
+	pinctrl_spdif_sleep: spdifgrpsleep {
+		fsl,pins = <
+			MX93_PAD_GPIO_IO22__GPIO2_IO22		0x31e
+			MX93_PAD_GPIO_IO23__GPIO2_IO23		0x31e
+		>;
+	};
+};
-- 
2.25.1

